### PR TITLE
Add ConsoleGui prototype

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
This adds a new module called `ConsoleGuiTools` which enables a xplat `Out-ConsoleGridView` right in a terminal using [gui.cs](https://github.com/migueldeicaza/gui.cs).

It supports passthru

image

### CODE REVIEW

File: src/Microsoft.PowerShell.ConsoleGuiTools/Microsoft.PowerShell.ConsoleGuiTools.csproj [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
+    <Import Project="..\..\GraphicalTools.Common.props" />
+    <PropertyGroup>
+        <TargetFramework>netcoreapp3.0</TargetFramework>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Nstack.Core" Version="0.12.0" />
+        <PackageReference Include="Terminal.Gui" Version="0.24.0" />
```
Comment (TylerLeonhardt) [inline src/Microsoft.PowerShell.ConsoleGuiTools/Microsoft.PowerShell.ConsoleGuiTools.csproj]: We need to wait for 0.25.0 which has fixes to `ListView`

Comment (TylerLeonhardt) [reply src/Microsoft.PowerShell.ConsoleGuiTools/Microsoft.PowerShell.ConsoleGuiTools.csproj]: Issue tracking release: https://github.com/migueldeicaza/gui.cs/issues/328

**Comment by corbob** [comment]
I think you mean for the link to go to https://github.com/migueldeicaza/gui.cs (it looks like you're missing the https so github is trying to create a pull request or a compare or something 🤷‍♂

File: src/Microsoft.PowerShell.ConsoleGuiTools/Microsoft.PowerShell.ConsoleGuiTools.csproj (hunk 1 above)
Comment (TylerLeonhardt) [reply src/Microsoft.PowerShell.ConsoleGuiTools/Microsoft.PowerShell.ConsoleGuiTools.csproj]: moved to latest Terminal.Gui!

File: src/Microsoft.PowerShell.ConsoleGuiTools/ConsoleGui.cs [hunk 2]
```diff
@@ -0,0 +1,137 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleGuiTools/ConsoleGui.cs]: The `All rights reserved.` is no longer part of the official header: https://docs.opensource.microsoft.com/content/releasing/copyright-headers.html

File: src/Microsoft.PowerShell.ConsoleGuiTools/ConsoleGui.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+using System.Runtime.InteropServices;
+using System.Text;
+using OutGridView.Models;
+using Terminal.Gui;
+
+namespace OutGridView.Cmdlet
+{
+    internal class ConsoleGui
+    {
+        private const string ACCEPT_TEXT = "Are you sure you want to select\nthese items to send down the pipeline?";
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleGuiTools/ConsoleGui.cs]: Doesn't have to be part of this PR, but this text should be in a resx for localization purposes.

File: src/Microsoft.PowerShell.ConsoleGuiTools/ConsoleGui.cs (hunk 2 above)
Comment (TylerLeonhardt) [reply src/Microsoft.PowerShell.ConsoleGuiTools/ConsoleGui.cs]: fixed.

File: docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+### System.Management.Automation.PSObject
+You can send any object to this cmdlet.
+
+## OUTPUTS
+
+### None
+**Out-ConsoleGridView** does not return any objects.
+
+## NOTES
+* You cannot use a remote command to open a grid view window on another computer.
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md]: copy paste... removed

File: docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+## INPUTS
+
+### System.Management.Automation.PSObject
+You can send any object to this cmdlet.
+
+## OUTPUTS
+
+### None
+**Out-ConsoleGridView** does not return any objects.
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md]: updated to what is in our [Out-GridView docs](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/7.0/Microsoft.PowerShell.Utility/Out-GridView.md#outputs)

File: docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+
+### -PassThru
+Indicates that the cmdlet sends items from the interactive window down the pipeline as input to other commands.
+By default, this cmdlet does not generate any output.
+This parameter is equivalent to using the Multiple value of the *OutputMode* parameter.
+
+To send items from the interactive window down the pipeline, click to select the items and then click OK.
+Shift-click and Ctrl-click are supported.
+
+This parameter was introduced in Windows PowerShell 3.0.
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md]: copy paste... removed.

File: docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+By default, this cmdlet does not generate any output.
+To send items from the interactive window down the pipeline, click to select the items and then click OK.
+
+The values of this parameter determine how many items you can send down the pipeline.
+
+- None.  No items. This is the default value.
+- Single.  Zero items or one item. Use this value when the next command can take only one input object.
+- Multiple.  Zero, one, or many items.  Use this value when the next command can take multiple input objects. This value is equivalent to the *Passthru* parameter.
+
+This parameter was introduced in Windows PowerShell 3.0.
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md]: copy paste... removed.

File: docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+As a result, the output from the Get-ChildItem command is saved in the $A variable before it is sent to **Out-ConsoleGridView**.
+
+### Example 5: Output processes for a specified computer to a grid view
+```
+PS C:\> Get-Process -ComputerName "Server01" | ogv -Title "Processes - Server01"
+```
+
+This command displays the processes that are running on the Server01 computer in a grid view window.
+
+The command uses `ogv`, which is the built-in alias for the **Out-ConsoleGridView** cmdlet, it uses the *Title* parameter to specify the window title.
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md]: the alias is `ocgv`. copy paste... fixed.

File: docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+Module Name: Microsoft.PowerShell.ConsoleGuiTools
+ms.date: 08/09/2019
+schema: 2.0.0
+title: Out-ConsoleGridView
+---
+
+# Out-ConsoleGridView
+
+## SYNOPSIS
+Sends output to an interactive table in a separate window.
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply docs/Microsoft.PowerShell.ConsoleGuiTools/Out-ConsoleGridView.md]: copy paste... fixed.

File: GraphicalTools.build.ps1 [hunk 10]
```diff
@@ -102,83 +86,104 @@ task SetupDotNet -Before Clean, Build {
...
+        }
+
+        foreach ($projName in $moduleLayout.NativeBuildAssets.Keys) {
+            exec { & $script:dotnetExe clean -c $Configuration "$PSScriptRoot/src/$projName/$projName.csproj" }
+        }
+    }
 
-    Get-ChildItem "$PSScriptRoot\module\$script:ModuleName\Commands\en-US\*-help.xml" -ErrorAction Ignore | Remove-Item -Force -ErrorAction Ignore
+    foreach ($mn in $ModuleName) {
+        Get-ChildItem "$PSScriptRoot\module\$mn\Commands\en-US\*-help.xml" -ErrorAction Ignore | Remove-Item -Force -ErrorAction Ignore
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply GraphicalTools.build.ps1]: removed.

File: GraphicalTools.build.ps1 [hunk 11]
```diff
@@ -1,36 +1,20 @@
...
+
+    [string[]]$ModuleName = @( "Microsoft.PowerShell.GraphicalTools", "Microsoft.PowerShell.ConsoleGuiTools" ) 
 )
 
 $script:IsUnix = $PSVersionTable.PSEdition -and $PSVersionTable.PSEdition -eq "Core" -and !$IsWindows
 
-$script:ModuleName = "Microsoft.PowerShell.GraphicalTools"
-$script:ModuleBinPath = "$PSScriptRoot/module/$script:ModuleName/"
 $script:TargetFramework = "netcoreapp3.0"
 $script:RequiredSdkVersion = "3.0.100-preview5-011568"
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply GraphicalTools.build.ps1]: switched to using `global.json`

**Pull request merged**